### ISSUE TITLE
[WK01-Day 002][21.08.03.Tue.] 확률론, 그리고 4번 문제

### ISSUE THREAD

**Issue by KimDaeUng** [Boostcamp-AI-Tech-2th-level01-04/TWL#2]
# Intro
다른 일도 여러가지 겹쳐 있어서 오늘은 최대한 학습을 빨리 끝내고 처리해야지 했는데,
과제 하나에 발목 잡혔다. 필수과제 4번 문제에만 4시간 반 동안 삽질했다.

# 강의 복습 내용
- 필수 퀴즈. 이미 다 아는 거니 금방 풀겠지~ 했는데 안 본 사이에 막상 문제 풀려니 가물가물 했다.
- 결국 [AI Math: 6강 확률론] 강의로 들어가 내용을 다시 정리하였고, 알고 있던 내용은 스킵하고, 개념과 개념간의 연결, 그래서 이 개념이 왜 중요하고 어떤 의미를 갖는지에 좀 더 집중했다.

## 딥러닝에서의 확률론
- 딥러닝의 바탕은 확률론 기반의 기계학습 이론 
- loss function의 작동원리는 데이터 공간을 통계적으로 해석해 유도된다.
  - L2-norm: 예측오차 분산을 가장 최소화하는 방향으로 학습하도록 유도
  - Cross-Entropy: 모델 예측의 불확실성 최소화하는 방향으로 학습하도록 유도
- 통계적 기계학습의 기본 원리: 예측이 틀릴 위험(risk)를 최소화하도록 데이터를 학습
- 이러한 분산 및 불확실성 최소화를 위해서는 측정 방법을 알아야한다. 이러한 방법을 통계학에서 제공하기 때문에 확률론의 기본 개념을 알아야함

## 조건부확률과 기계학습
- 딥러닝 모델은 데이터 $\bold{x}$로부터 패턴 $\pi$을 추출하는 함수
- 로지스틱 회귀의 선형모델과 소프트맥스 함수의 결합($g(z) = softmax(\bold{W}z + \bold{b})$) : 함수의 입력값을 기반으로 확률로 변환(입력 패턴을 변환해 카테고리 분포를 모델링하는 역할)
- 분류 문제에서는 데이터 x로부터 추출된 특징 패턴 $\pi(\bold{x})$을 앞의 $g(x)$와 결합하여 입력 데이터 $\bold{x}$에 대한 조건부 확률  $P(y|\bold{x})$를 계산하는 것
- 회귀 문제에서는 조건부 기대값 $E[y|\bold{x}]$를 추정
- 특징패턴을 추출하는 함수 $\pi$의 파라미터(모수)를 어떻게 찾을 것인가(손실함수의 설정 문제)는 기계학습 문제와 모델에 따라 달라짐

## 몬테카를로 샘픞링
- 확률 분포를 모를 때(대부분의 기계학습 문제가 이에 해당), 데이터를 사용해 기대값을 계산하는 방법
- 이산형, 연속형에 관계없이 성립

# 과제 수행 과정 / 결과물 
- 개별 함수의 구현은 어렵지 않았으나 main 함수의 출력양식과 관련된 테스트 케이스를 통과하지 못했다.
- 사용자의 입력을 받고, 종료 커맨드를 입력하기 전까지 반복적으로 작동하는 프로그램을 작성하려면 while문으로 전체 코드를 감싼다.
- 사용자 입력이 잘못 입력되었을 때 즉시 다시 입력을 받도록 하기 위해 내부 while 문 사용
- 사용자 입력을 받는 코드를 두 번이나 작성했는데 더 깔끔한 방법이 있을 것 같다(`if while True:`로 처리후 내부에 `user_input = ...`을 옮기는 방법)
-  0입력시 종료조건 코드는 최초 숫자 입력시 필요한 것인지 모르겠다. strke를 얻고 난 후의 사용자 입력시 조건으로만 부여되어 있는데, 최초 숫자 입력시에는 저 조건을 넣지 않아도 작동한다.
- 문제에서 명확하게 주어지지 않아 헷갈린 조건들, 각 조건들을 바꿔가며 시행착오를 거쳤으나 결국 출력 양식을 서치하여(솔루션은 존재하지 않았습니다). 이에 맞게 수정하였다.
  - Random number를 사용자가 볼 수 있도록 입력 받기 전에 표시해야하는지, 아니면 보지 못하게 받은 후에 표시해야하는지
  - 숫자를 잘못 입력했을 떄, 컴퓨터의 Random를 새로 부여하는지
  - Strike 와 Ball의 출력 양식
  - 재 시작시  "Play Baseball" 메시지를 출력하는지 아니면 최초 실행시에만 출력하는지

```python
user_input = input('Input guess number : ')

while not is_validated_number(user_input):
    if user_input == '0':
        break
    print('Wrong Input')
    user_input = input('Input guess number : ')
  
if user_input == '0':
    break
```

# 피어세션
- 피어세션이 피어씁니다. 발표 준비
- 팀 구글 드라이브 개설
- 그라운드 룰 업데이트

# 학습 회고
- 슬랙과 팀 단톡 모두 4번 문제에 대한 질문이 자주 보였다. 질문을 주고 받으면서 문제 풀이에 대한 힌트를 얻을 수 있어 도움이 되었다.
- 다 안다고 방심했던 부분에서 발목잡혔다. 대충하지 말고 확실하게 하자.
